plugin/pkg/client/auth/oidc/oidc.go

```diff
@@ -0,0 +1,168 @@
+package oidc
+
+import (
+	"encoding/base64"
+	"errors"
+	"fmt"
+	"net/http"
+	"strings"
+	"time"
+
+	"github.com/coreos/go-oidc/jose"
+	"github.com/coreos/go-oidc/oauth2"
+	"github.com/coreos/go-oidc/oidc"
+	"github.com/golang/glog"
+
+	"k8s.io/kubernetes/pkg/client/restclient"
+)
+
+func init() {
+	if err := restclient.RegisterAuthProviderPlugin("oidc", newOIDCAuthProvider); err != nil {
+		glog.Fatalf("Failed to register oidc auth plugin: %v", err)
+	}
+}
+
+func newOIDCAuthProvider(_ string, cfg map[string]string, persister restclient.AuthProviderConfigPersister) (restclient.AuthProvider, error) {
+	var certAuthData []byte
+	var err error
+	if cfg["idp-certificate-authority-data"] != "" {
+		certAuthData, err = base64.StdEncoding.DecodeString(cfg["idp-certificate-authority-data"])
+		if err != nil {
+			return nil, err
+		}
+	}
+
+	clientConfig := restclient.Config{
+		TLSClientConfig: restclient.TLSClientConfig{
+			CAFile: cfg["idp-certifiate-authority"],
+			CAData: certAuthData,
+		},
+	}
+
+	trans, err := restclient.TransportFor(&clientConfig)
+	if err != nil {
+		return nil, err
+	}
+	hc := &http.Client{Transport: trans}
+
+	issuer, ok := cfg["idp-issuer-url"]
+	if !ok || issuer == "" {
+		return nil, errors.New("Must provide idp-issuer-url")
+	}
+
+	providerCfg, err := oidc.FetchProviderConfig(hc, strings.TrimSuffix(issuer, "/"))
+	if err != nil {
+		return nil, fmt.Errorf("error fetching provider config: %v", err)
+	}
+
+	scopes := strings.Split(cfg["extra-scopes"], ",")
```

Comment:
https://godoc.org/github.com/coreos/go-oidc/oidc#ClientConfig doesn't document what the `scopes` field of `oidc.ClientConfig` does.  In the [OIDC documentation](http://openid.net/specs/openid-connect-core-1_0.html#ScopeClaims), the scopes mentioned are `email`,`profile`,`address`, and `phone`.  The specification calls `azp` a "claim".